rotary.py

```diff
@@ -7,107 +7,137 @@
+    # CLK/DT    CLK/DT     CLK/DT    CLK/DT
+    #   00        01         10        11
+    [_R_START, _R_CCW_1, _R_CW_1,  _R_START],             # _R_START
+    [_R_CW_2,  _R_START, _R_CW_1,  _R_START],             # _R_CW_1
+    [_R_CW_2,  _R_CW_3,  _R_CW_1,  _R_START],             # _R_CW_2
+    [_R_CW_2,  _R_CW_3,  _R_START, _R_START | _DIR_CW],   # _R_CW_3
+    [_R_CCW_2, _R_CCW_1, _R_START, _R_START],             # _R_CCW_1
+    [_R_CCW_2, _R_CCW_1, _R_CCW_3, _R_START],             # _R_CCW_2
+    [_R_CCW_2, _R_START, _R_CCW_3, _R_START | _DIR_CCW],  # _R_CCW_3
+    [_R_START, _R_START, _R_START, _R_START]]             # _R_ILLEGAL
 
 _STATE_MASK = const(0x07)
-_DIR_MASK = const(0x30)  
+_DIR_MASK = const(0x30)
+
 
 def _wrap(value, incr, lower_bound, upper_bound):
     range = upper_bound - lower_bound + 1
-    value = value + incr    
-    
+    value = value + incr
+
     if value < lower_bound:
         value += range * ((lower_bound - value) // range + 1)
-    
-    return lower_bound + (value - lower_bound) % range     
+
+    return lower_bound + (value - lower_bound) % range
+
 
 def _bound(value, incr, lower_bound, upper_bound):
     return min(upper_bound, max(lower_bound, value + incr))
 
-class Rotary(object): 
- 
-    RANGE_UNBOUNDED = const(1)  
+
+def _trigger(rotary_instance):
+    for listener in rotary_instance._listener:
+        listener()
+
+class Rotary(object):
+
+    RANGE_UNBOUNDED = const(1)
     RANGE_WRAP = const(2)
     RANGE_BOUNDED = const(3)
- 
+
     def __init__(self, min_val, max_val, reverse, range_mode):
         self._min_val = min_val
         self._max_val = max_val
         self._reverse = -1 if reverse else 1
         self._range_mode = range_mode
         self._value = min_val
         self._state = _R_START
-        
-    def set(self, value=None, min_val=None, max_val=None, reverse=None, range_mode=None):
+        self._listener = []
+
+    def set(self, value=None, min_val=None,
+            max_val=None, reverse=None, range_mode=None):
         # disable DT and CLK pin interrupts
         self._hal_disable_irq()
-        
-        if value != None:
+
+        if value is not None:
             self._value = value
-        if min_val != None:
+        if min_val is not None:
             self._min_val = min_val
-        if max_val != None:
+        if max_val is not None:
             self._max_val = max_val
-        if reverse != None:
+        if reverse is not None:
             self._reverse = -1 if reverse else 1
-        if range_mode != None:
+        if range_mode is not None:
             self._range_mode = range_mode
         self._state = _R_START
 
         # enable DT and CLK pin interrupts
         self._hal_enable_irq()
-    
+
     def value(self):
         return self._value
 
     def reset(self):
         self._value = 0
-        
+
     def close(self):
         self._hal_close()
-        
+
+    def add_listener(self, l):
+        self._listener.append(l)
```

Comment:
What do you think about adding a `remove_listener()` method?

Reply:
that's a good idea.